### ISSUE TITLE
ENH: expose sindex.nearest

### ISSUE THREAD

**Issue by martinfleis** [geopandas/geopandas#1977]
We should expose `sindex.nearest` in the same way we expose `query_bulk` etc. It essentially means only to wrap the method on our own and ensure that shapely geometries are converted to pygeos geometries.

See https://github.com/pygeos/pygeos/issues/363 for the context.

RTree can just return `NotImplementedError`.

**Comment by tomalrussell** [comment]
I've made an attempt at this over in #2053 - I hope that looks like it's along the right lines.

For the RTree case, in the past I've accessed the undocumented sindex directly, doing something like this, in [`snkit.network#L457-L465`](https://github.com/tomalrussell/snkit/blob/167d72ee944977e7e8d53dffd2979f9cfee30f0f/src/snkit/network.py#L457-L465):
```python
def nearest(geom, gdf):
    """Find the element of a GeoDataFrame nearest a shapely geometry
    (This assumes `gdf.sindex` is an `rtree.index.Index`)
    """
    matches_idx = gdf.sindex.nearest(geom.bounds)
    nearest_geom = min(
        [gdf.iloc[match_idx] for match_idx in matches_idx],
        key=lambda match: geom.distance(match.geometry)
    )
    return nearest_geom
```

Which should be neatly simplified to something like this, if I've understood it all right:
```python
def nearest(geom, gdf):
    """Find the element of a GeoDataFrame nearest a shapely geometry
    (This assumes `gdf.sindex.nearest` is a thin wrapper around `pygeos.STRTree.nearest`)
    """
    match_idx = gdf.sindex.nearest(geom)
    return gdf.iloc[match_idx]
```

I'm happy to leave RTree unimplemented if it's on the way to deprecation anyway - but could equally take a look at that implementation too.

**Comment by jorisvandenbossche** [comment]
The `nearest` method was added in https://github.com/geopandas/geopandas/pull/2053, and the `nearest_all` in https://github.com/geopandas/geopandas/pull/2053.

There is one more aspect related to the API that I wanted to bring up, though. In the `sjoin_nearest` PR, @brendan-ward brought up the following point (https://github.com/geopandas/geopandas/pull/1865#pullrequestreview-733447923):

> In relation to your `sindex.nearest_all` here and `sindex.nearest` in #2053, I'm wondering if it would be best to use `STRtree.nearest_all` for `sindex.nearest` rather than exposing both functions from pygeos here, but I don't have strong opinions about it - mostly that it seems easier to communicate about a single top-level nearest function.

to which @martinfleis answered (https://github.com/geopandas/geopandas/pull/1865#issuecomment-901763108):

> I wouldn't deviate from pygeos API unless there's a really good reason to do so. In this case I don't see any issue in exposing both `nearest` and `nearest_all` in our API.

Although I certainly agree that ideally we keep it consistent with pygeos (although we also have discussed that our sindex SpatialIndex API can also be intended to wrap other spatial index implementations), there might be some reason to reconsider this:

1) The interface of pygeos _might_ still have to change when integrating it with Shapely, given the existing interface there. Shapely currently already has a `nearest` method (which returns a single geometry), and added a `nearest_geom` and `nearest_item` in the upcoming release. But those both still return a single geometry or single index. While the `nearest`/`nearest_all` from pygeos is more like a "nearest bulk" (in terms of `query_bulk`, returning an array of left+right indices).

2) Personally, I have found the `nearest`/`nearest_all` somewhat confusing. I think a keyword in a single nearest method to indicate whether to return all equidistant or only a single might be nicer (but this is certainly subjective). In the end, both methods have an identical return type, so are also easy to integrate (I understand that from pygeos' point of view the implementation is quite different, but from the user point of view, this feels more like an option)

cc @adriangb

**Comment by jorisvandenbossche** [comment]
I also opened https://github.com/pygeos/pygeos/issues/398 on the pygeos side about those inconsistencies of the PyGEOS API with future Shapely 2.0 API.

**Comment by brendan-ward** [comment]
> nearest/nearest_all somewhat confusing

In case it helps here, I've been using `nearest_all` quite a bit since it came out in pygeos - but the *outstanding advantage* of it over the plain `nearest` is that I can set a maximum search distance.  That it can return multiple equidistant nearest neighbors for each input really only shows up for me when there are intersecting geometries (haven't seen a case in practice of non-intersecting equidistant neighbors in my analyses yet), and I think those are a special enough case that doing analysis outside of nearest is necessary (i.e., if I don't care about any other criteria, just take the first nearest neighbor; if I want to filter out intersections to do a `query_bulk` first and exclude those from `nearest_all`; or some other criteria like of the intersecting ones which is closest to the centroid of a poly, etc).

The potential disadvantage of `nearest_all` is that it does quite a bit more work internally within pygeos to realize this - but that extra work may be more than paid for by the efficiences of not searching against widely distant neighbors, and the secondary advantage of not having to then filter nearest neighbors by a maximum distance afterward.

It is possible to imagine a future where there is a distance-constrained nearest neighbor GEOS CAPI function, since I think adding a distance constraint there could be straightforward, but it seems less likely that the CAPI would be further expanded to return all equidistant nearest neighbors rather than an arbitrary first one like it does now.  In which case, having `nearest` type functionality but with an optional distance constraint via a keyword could give us a public API here that lets us ease into this?  (internally would use either pygeos `nearest` or `nearest_all`).

Perhaps another way to look at this though: we want a public API here that is simple and clear.  For power users that need more advanced functionality, they can use some of the internals or pygeos functions more directly.  As I see it, pygeos gives us the building blocks to compose into simpler higher order APIs here; we don't _have_ to mimic the API from pygeos here, though I agree with @martinfleis that as a general rule following the underlying APIs is a pragmatic choice.  Taking the 80% / 20% approach, is there an API here that covers 80% of the uses that we can implement, and delegate the other 20% to instructing users to use the underlying libraries / internals directly?

**Comment by jorisvandenbossche** [comment]
We discussed this on the community call today, and our idea now is to combine both in a single method, with a keyword switching between both behaviours. When `max_distance=None` and you don't want all equi-distant nearest geometries, it can under the hood use pygeos' `nearest`, and otherwise `nearest_all` (and this makes we can also support max_distance with only returning a single nearest geometry).

I opened a PR with this change at https://github.com/geopandas/geopandas/pull/2147

**Comment by jorisvandenbossche** [comment]
Finally closed by #2147
